src/engine/sync/synccontrol.cpp

```diff
@@ -297,32 +315,45 @@ void SyncControl::setInstantaneousBpm(double bpm) {
 void SyncControl::reportTrackPosition(double fractionalPlaypos) {
     // If we're close to the end, and master, disable master so we don't stop
     // the party.
-    if (getSyncMode() == SYNC_MASTER &&
+    if (isMaster(getSyncMode()) &&
             fractionalPlaypos > kTrackPositionMasterHandoff) {
-        m_pChannel->getEngineBuffer()->requestSyncMode(SYNC_NONE);
+        m_pChannel->getEngineBuffer()->requestSyncMode(SYNC_FOLLOWER);
     }
 }
 
+// called from an engine worker thread
 void SyncControl::trackLoaded(TrackPointer pNewTrack) {
-    //qDebug() << getGroup() << "SyncControl::trackLoaded";
-    if (getSyncMode() == SYNC_MASTER) {
-        // If we change or remove a new track while master, hand off.
-        m_pChannel->getEngineBuffer()->requestSyncMode(SYNC_NONE);
-    }
+    BeatsPointer pBeats;
     if (pNewTrack) {
-        m_masterBpmAdjustFactor = kBpmUnity;
-        if (isSynchronized()) {
-            // Because of the order signals get processed, the file/local_bpm COs and
-            // rate slider are not updated as soon as we need them, so do that now.
-            m_pFileBpm->set(pNewTrack->getBpm());
-            m_pLocalBpm->set(pNewTrack->getBpm());
-            // We used to set the m_pBpm here, but that causes a signal loop whereby
-            // that was interpreted as a rate slider tweak, and the master bpm
-            // was changed.  Instead, now we pass the suggested bpm to enginesync
-            // explicitly, and it can decide what to do with it.
-            m_pEngineSync->notifyTrackLoaded(this, m_pLocalBpm->get() * m_pRateRatio->get());
-        }
+        pBeats = pNewTrack->getBeats();
     }
+    trackBeatsUpdated(pBeats);
+}
+
+void SyncControl::trackBeatsUpdated(BeatsPointer pBeats) {
+    // This slot is fired by a new file is loaded or if the user
+    // has adjusted the beatgrid.
+    if (kLogger.traceEnabled()) {
+        kLogger.trace() << getGroup() << "SyncControl::trackBeatsUpdated";
+    }
+
+    VERIFY_OR_DEBUG_ASSERT(m_pLocalBpm) {
+        // object not initalized
+        return;
+    }
+
+    m_pBeats = pBeats;
+    m_masterBpmAdjustFactor = kBpmUnity;
+
+    SyncMode syncMode = getSyncMode();
+    if (syncMode == SYNC_MASTER_SOFT) {
+        // If we change or remove beats while soft master, hand off.
+        m_pChannel->getEngineBuffer()->requestSyncMode(SYNC_FOLLOWER);
+    } else if (syncMode == SYNC_FOLLOWER) {
+        // Refresh follower state to update soft master
+        m_pChannel->getEngineBuffer()->requestSyncMode(SYNC_FOLLOWER);
+    }
```

Comment:
How about an else branch? Moreover both if and else if branches contain the same code.

Reply:
no need for an else branch here -- nothing to do if sync is explicit or none.  But I removed the redundant code